src/gradual_utils.erl

```diff
@@ -0,0 +1,299 @@
+-module(gradual_utils).
+
+-include("log.hrl").
+
+-export([
+    init/0,
+    clean/0,
+    preprocess_constrs/1,
+    postprocess/3,
+    subst_ty/3
+]).
+
+-ifdef(TEST).
+-export([
+  collect_pos_neg_tyvars/1,
+  replace_dynamic/1
+]).
+-endif.
+
+-define(VAR_ETS, framevariable_counter_ets_table).
+-define(ALL_ETS, [?VAR_ETS]).
+
+
+-spec init() -> _.
+init() ->
+  case ets:whereis(?VAR_ETS) of
+      undefined -> 
+        [ets:new(T, [public, set, named_table]) || T <- ?ALL_ETS],
+        ets:insert(?VAR_ETS, {variable_id, 0});
+      _ -> ok
+  end.
+
+-spec clean() -> _.
+clean() ->
+  case ets:whereis(?VAR_ETS) of
+      undefined -> ok;
+      _ -> 
+        [ets:delete(T) || T <- ?ALL_ETS]
+  end.
+
+
+-spec fresh_typevar_name() -> ast:ty_varname().
+fresh_typevar_name() ->
+    NewId = ets:update_counter(?VAR_ETS, variable_id, {2,1}),
```

Comment:
This is global state outside the `erlang_types` project. Better to use the context for variable counters.